rescript/ncbi.py

```diff
@@ -88,87 +114,216 @@ def get_ncbi_data(
-    if ids:
-        assert len(ids) >= 1, "need at least one id"
-        epost = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi'
-        data = {'db': params['db'], 'id': ','.join(ids)}
+def _get_logger(logging_level):
+    logger = logging.getLogger('rescript')
+    if not hasattr(logger, 'isSetUp'):
+        formatter = logging.Formatter(
+            '%(levelname)s:%(asctime)-15s:%(processName)s:%(message)s')
+        handler = logging.StreamHandler(sys.stdout)
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+        if logging_level:
+            logger.setLevel(logging_level)
+        logger.isSetUp = True
+    return logger
+
+
+def _robustify(http_request, logger, *args):
+    max_retries = 10
+    backoff_factor = 1
+    max_backoff = 120
+    status_forcelist = [400, 429, 500, 502, 503, 504]
+    exception_forcelist = (
+        ChunkedEncodingError, ConnectionError, ExpatError, ReadTimeout)
+    for retry in range(max_retries):
+        try:
+            return http_request(*args)
+        except HTTPError as e:
+            if e.response.status_code == 400:
+                logger.debug(
+                    'Request failed with code 400. This could be because all '
+                    'of the requested accession ids are invalid, or it could '
+                    'just be a temporary failure. Retrying.')
+            elif e.response.status_code in status_forcelist:
+                logger.debug('Request failed with code '
+                             + str(e.response.status_code) + '. Retrying.')
+            else:
+                raise e
+            last_exception = e
+        except RuntimeError as e:
+            if str(e) == 'bad record':
+                logger.debug('Request failed. Retrying.')
+            else:
+                raise e
+            last_exception = e
+        except exception_forcelist as e:
+            logger.debug('Request failed with exception\n' +
+                         type(e).__name__ + ': ' + str(e) + '\nRetrying.')
+            last_exception = e
+        time.sleep(min(backoff_factor*2**retry, max_backoff))
+    raise RuntimeError(
+        'Maximum retries (10) exceeded for HTTP request. Persistent trouble '
+        'downloading from NCBI. Last exception was\n' +
+        type(last_exception).__name__ + ': ' + str(last_exception))
+
+
+def _epost(params, ids, request_lock, logging_level, entrez_delay=0.334):
+    assert len(ids) >= 1, "need at least one id"
+    epost = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi'
+    data = {'db': params['db'], 'id': ','.join(ids)}
+    logger = _get_logger(logging_level)
+
+    def request(params):
+        request_lock.acquire()
+        logger.debug('request lock acquired')
         time.sleep(entrez_delay)
-        r = requests.post(epost, data=data)
+        try:
+            r = requests.post(
+                epost, data=data, params=_entrez_params, timeout=10,
+                stream=True)
```

Comment:
Thanks @misialq, I tried that but it didn't work first time and it wasn't important to me so I left it as it is.

The point of setting `stream=True` is because NCBI doesn't care how many connections you've got downloading at any given moment, it only cares about how many connections you make per second. So making the post goes inside the lock with a delay to prevent too many connections being made per second, but then the lock can be released while the download proceeds.

Reply:
Alright, thanks for the explanation!